SBOM issue is still there there is successful pipeline AI.Models.NL2Flow.MultiTurn Repo

Hello Team,

Though there is successful official pipeline, still showing SBOM issues for the repo “AI.Models.NL2Flow.MultiTurn-Official”.

https://dev.azure.com/powerbi/Business360%20AI/_build/results?buildId=8197256&view=logs&j=0029b136-61e5-522f-ecc4-230ce6a150b5&t=b4ad4c61-af02-5b5d-29c6-f4b9746ae1f5](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdev.azure.com%2Fpowerbi%2FBusiness360%2520AI%2F_build%2Fresults%3FbuildId%3D8197256%26view%3Dlogs%26j%3D0029b136-61e5-522f-ecc4-230ce6a150b5%26t%3Db4ad4c61-af02-5b5d-29c6-f4b9746ae1f5&data=05%7C01%7Chravella%40microsoft.com%7C4817b1916f944d7c419008dbce53d1be%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638330631649125611%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=SVDfGBdEUKkpzen1Goqglv9DMehb3%2FkXXs1MA5fWqhs%3D&reserved=0)

https://dev.azure.com/powerbi/Business360%20AI/_build?definitionId=9290&_a=summary 
 
there are couple of below error messages in logs.
 
##[information]Getting Python data from [https://pypi.org/pypi/AI.Models/json](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fpypi.org%2Fpypi%2FAI.Models%2Fjson&data=05%7C01%7Chravella%40microsoft.com%7C4817b1916f944d7c419008dbce53d1be%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638330631649132235%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=sS3vyoyvAO3y4bWPCRiNkxp%2FDea8m%2BIDVpaPKN9VBZk%3D&reserved=0)
##[error]Component "0.0.3" : {ReleaseValue)} could not be added to the sorted list of pip components for spec="AI.Models". Usually this happens with unexpected PyPi version formats (e.g. prerelease/dev versions).
System.ArgumentException: An item with the same key has already been added. Key: 0.0.3
at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
at Microsoft.ComponentDetection.Detectors.Pip.PythonVersion.Create(String version)
at Microsoft.ComponentDetection.Detectors.Pip.PyPiClient.GetReleasesAsync(PipDependencySpecification spec)
##[error]Component "0.1.3" : {ReleaseValue)} could not be added to the sorted list of pip components for spec="AI.Models". Usually this happens with unexpected PyPi version formats (e.g. prerelease/dev versions).
System.ArgumentException: An item with the same key has already been added. Key: 0.1.3
at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
at Microsoft.ComponentDetection.Detectors.Pip.PythonVersion.Create(String version)
at Microsoft.ComponentDetection.Detectors.Pip.PyPiClient.GetReleasesAsync(PipDependencySpecification spec)
##[warning]Root dependency "AI.Models" not found on pypi. Skipping package.
##[warning]Root dependency "AI.Models" not found on pypi. Skipping package.
##[information]Getting Python data from [https://pypi.org/pypi/torch/json](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fpypi.org%2Fpypi%2Ftorch%2Fjson&data=05%7C01%7Chravella%40microsoft.com%7C4817b1916f944d7c419008dbce53d1be%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638330631649137682%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=1Ac9T2cj8rKN4T%2F9uOOl7zSWkOTFt6%2BK6joNFnzFkiQ%3D&reserved=0)
##[information]Getting Python data from [https://pypi.org/pypi/torch/json](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fpypi.org%2Fpypi%2Ftorch%2Fjson&data=05%7C01%7Chravella%40microsoft.com%7C4817b1916f944d7c419008dbce53d1be%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638330631649143075%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=nfFRDCfFz9WKl9uJb1GCXD%2F2FE7%2FWfrI%2F22y8e9Ql7s%3D&reserved=0)
##[information]Getting Python data from [https://pypi.org/pypi/pyspark/json](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fpypi.org%2Fpypi%2Fpyspark%2Fjson&data=05%7C01%7Chravella%40microsoft.com%7C4817b1916f944d7c419008dbce53d1be%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638330631649148392%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=oHwD96br5%2BMUQdJL0XKMebJ%2BxwRyBSrQbk%2BQphj1Tng%3D&reserved=0)
##[information]Getting Python data from [https://pypi.org/pypi/pyspark/json](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fpypi.org%2Fpypi%2Fpyspark%2Fjson&data=05%7C01%7Chravella%40microsoft.com%7C4817b1916f944d7c419008dbce53d1be%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638330631649153689%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=NM51hNAuf0yN%2BR60o%2B0BikESfo4BoR7NQOx163xz3bA%3D&reserved=0)
##[error]Component "3.0.2" : {ReleaseValue)} could not be added to the sorted list of pip components for spec="pyspark". Usually this happens with unexpected PyPi version formats (e.g. prerelease/dev versions).
System.ArgumentException: An item with the same key has already been added. Key: 3.0.2
at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
at Microsoft.ComponentDetection.Detectors.Pip.PythonVersion.Create(String version)
at Microsoft.ComponentDetection.Detectors.Pip.PyPiClient.GetReleasesAsync(PipDependencySpecification spec)
##[error]Component "3.1.3" : {ReleaseValue)} could not be added to the sorted list of pip components for spec="pyspark". Usually this happens with unexpected PyPi version formats (e.g. prerelease/dev versions).
System.ArgumentException: An item with the same key has already been added. Key: 3.1.3

 
Can you please share your views, why this behaviour for the Repo “AI.Models.NL2Flow.MultiTurn-Official”. ?
 
Regards,
Hari

--- a/src/Microsoft.ComponentDetection.Detectors/pip/IPyPiClient.cs
+++ b/src/Microsoft.ComponentDetection.Detectors/pip/IPyPiClient.cs
@@ -215,7 +215,7 @@ public sealed class PyPiClient : IPyPiClient, IDisposable
             {
                 this.logger.LogError(
                     ae,
-                    "Component {ReleaseKey} : {ReleaseValue)} could not be added to the sorted list of pip components for spec={SpecName}. Usually this happens with unexpected PyPi version formats (e.g. prerelease/dev versions).",
+                    "Component {ReleaseKey} : {ReleaseValue} could not be added to the sorted list of pip components for spec={SpecName}. Usually this happens with unexpected PyPi version formats (e.g. prerelease/dev versions).",
                     release.Key,
                     JsonConvert.SerializeObject(release.Value),
                     spec.Name);

--- a/src/Microsoft.ComponentDetection.Detectors/pip/PythonVersion.cs
+++ b/src/Microsoft.ComponentDetection.Detectors/pip/PythonVersion.cs
@@ -1,6 +1,7 @@
 namespace Microsoft.ComponentDetection.Detectors.Pip;
 
 using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Diagnostics.CodeAnalysis;
 using System.Linq;
@@ -15,7 +16,7 @@ public class PythonVersion : IComparable<PythonVersion>
 
     private static readonly Dictionary<string, int> PreReleaseMapping = new Dictionary<string, int> { { "a", 0 }, { "alpha", 0 }, { "b", 1 }, { "beta", 1 }, { "c", 2 }, { "rc", 2 }, { "pre", 2 }, { "preview", 2 } };
 
-    private static readonly Dictionary<string, PythonVersion> Cache = new();
+    private static readonly ConcurrentDictionary<string, PythonVersion> Cache = new();
 
     private readonly Match match;
 
@@ -121,7 +122,11 @@ public class PythonVersion : IComparable<PythonVersion>
         }
 
         var pythonVersion = new PythonVersion(version);
-        Cache.Add(version, pythonVersion);
+
+        // Multiple API call threads can be running at once, if the same version
+        // is parsed twice use the last entry since they will be the same.
+        Cache[version] = pythonVersion;
+
         return pythonVersion;
     }
 

